src/net/db.rs

```diff
@@ -1325,14 +1331,53 @@ impl PeerDB {
     /// -- always include all allowed neighbors
     /// -- never include denied neighbors
     /// -- for neighbors that are neither allowed nor denied, sample them randomly as long as they're fresh.
-    pub fn get_initial_neighbors(
+    pub fn get_random_initial_neighbors(
         conn: &DBConn,
         network_id: u32,
         network_epoch: u8,
+        peer_version: u32,
         count: u32,
         block_height: u64,
     ) -> Result<Vec<Neighbor>, db_error> {
-        PeerDB::get_random_neighbors(conn, network_id, network_epoch, count, block_height, true)
+        PeerDB::get_random_neighbors(
+            conn,
+            network_id,
+            network_epoch,
+            peer_version,
+            count,
+            block_height,
+            true,
+        )
+    }
+
+    pub fn get_valid_initial_neighbors(
+        conn: &DBConn,
+        network_id: u32,
+        curr_network_epoch: u8,
+        peer_version: u32,
+        burn_block_height: u64,
+    ) -> Result<Vec<Neighbor>, db_error> {
+        // UTC time
+        let now_secs = util::get_epoch_time_secs();
+
+        // the peer_version check mirrors the check in `has_acceptable_epoch`:
+        //    (my_epoch <= peer_epoch) OR (curr_epoch <= peer_epoch)
+        let query = "SELECT * FROM frontier WHERE initial = 1 AND (allowed < 0 OR ?1 < allowed) \
+         AND network_id = ?2 AND denied < ?3 AND ?4 < expire_block_height \
+         AND (?5 <= (peer_version & 0x000000ff) OR ?6 <= (peer_version & 0x000000ff))"
```

Comment:
Just flagging this here too -- I think you need `((?5 & 0x000000ff) <= (peer_version & 0x000000ff))`.